### PR TITLE
new mobile login page

### PR DESCRIPTION
**Pull request by czue**
First pass integration at new login page / styles

image

First commit just extracts some code that needed to be shared from a template and second is basically porting [this code](https://github.com/dimagi/mobile-dashboard-prototype/blob/master/app/templates/mobile-login.directive.html) into HQ.

### CODE REVIEW

File: custom/icds_reports/templates/icds_reports/mobile_login.html
Comment (czue) [inline custom/icds_reports/templates/icds_reports/mobile_login.html]: Mostly copied from https://github.com/dimagi/mobile-dashboard-prototype/blob/master/app/templates/mobile-login.directive.html

File: custom/icds_reports/static/icds_reports/mobile/css/app.css
Comment (czue) [inline custom/icds_reports/static/icds_reports/mobile/css/app.css]: Copied verbatim from https://github.com/dimagi/mobile-dashboard-prototype/blob/master/app/css/app.css

File: custom/icds_reports/templates/icds_reports/mobile_login.html
Comment (orangejenny) [inline custom/icds_reports/templates/icds_reports/mobile_login.html]: This isn't going to work because the initial page data infrastructure isn't included. Javascript translations also aren't going to work because `statici18n` isn't included. Take a look at the [SMS chat page](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/sms/templates/sms/chat.html) as an example of a page that doesn't inherit from `hqwebapp/base.html`: I think you'll want to add `get_current_language`, the `LANGUAGE_CODE` references, `statici18n`, `javascript_libraries`, `initial_page_data.js`, the `initial-page-data` div and likely the `commcarehq-urls` div.

**Review by orangejenny — CHANGES_REQUESTED** [review]

File: custom/icds_reports/templates/icds_reports/mobile_login.html
Comment (czue) [reply custom/icds_reports/templates/icds_reports/mobile_login.html]: Thanks, super helpful. I've updated this to add a minimal base template and used that in a few places.

**Comment by orangejenny** [comment]
@czue I like the approach of a minimal base template but I don't think it makes sense for it to be used by SMS chat and the ICDS mobile dashboard but not by any of the other base-type pages (hqwebapp/base.html, hqwebapp/base_mobile.html, styleguide/base.html, formplayer-common/base.html). I think ideally we'd have an even more minimal template that all of these derive from, probably like
```
hqwebapp/base_minimal.html
    hqwebapp/base_mobile.html
        icds_reports/mobile_login.html
        mobile/commcare_user_self_register.html
    hqwebapp/base_web.html (any page that uses HQ chrome, e.g., the top nav)
        ... most other pages in hq ...
    sms/chat.html
    formplayer-common/base.html
```

More practically, `javascript_libraries` isn't meant to be used multiple times in the same template. Check out the [source](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/templates/hqwebapp/includes/core_libraries.html). As-is, bootstrap and purify are getting included twice in `mobile_dashboard.html`. Reading through the SMS chat template again, I think it uses a more limited set of js than you'll want for the dashboard (apologies for the misdirection). You could either change chat.html to use `javascript_libraries` with `hq=True` (it would add some libraries that aren't used, but it's hardly the worst example of bloat in HQ) or not share code between the chat page and this mobile dashboard.

It's also possible that these various base templates don't fit well into a shared hierarchy and the most practical thing to do is just add some more partial templates (like grouping the initial-page-data, commcarehq-urls, and initial-analytics-data divs together) so that the amount of code copied between them is reduced.

**Comment by czue** [comment]
@orangejenny thanks for the great feedback and sorry I'm doing a bit of pattern-matching and forcing you to explain things to me that I could figure out if I spent more time understanding the internals of all this infrastructure. Hard to balance the desire to make progress with the complexity of all these subsystems.

I like both of your proposed approaches and also they are more than I'm comfortable biting off in terms of the QA footprint on the timelines we need to be moving for mobile dashboard (I'd really like to avoid touching `hqwebapp/base.html`). So I'd rather do something that has known scope of potential regression that is relatively easy to isolate and test.

So, I think practical options that would be feasible for the project timelines:

1. Create a custom base template specifically for ICDS mobile reports that addresses some of the concrete concerns you've raised (library import duplication and such - probably the simplest form being allowing sub-templates to override all of `javascript_libraries` with a `block` tag... and/or leave it out entirely). And assume all this code will stay custom to ICDS.
2. Take an approach like this one, with the same necessary cleanups and an overall plan to introduce a true inheritance chain, but leave that work for future maintainers to implement.

I'm leaning towards 1 but also happy with 2 if you think it's better. Thinking to our good old [code review guidelines](https://google.github.io/eng-practices/review/reviewer/standard.html) it's unclear to me whether option 2 would be "in a state where it definitely improves the overall code health of the system being worked on" and would defer to your knowledge on that.

**Comment by orangejenny** [comment]
@czue That all sounds reasonable. I agree changing the base template hierarchy introduces risk beyond the scope of this new dashboard. And I also agree with favoring your approach 1. It sounds like 1 means there will be some duplicated code within `custom/icds`, but I think it will be easier for a future dev to clean that up than to finish partially-implemented changes to the template hierarchy.

**Comment by czue** [comment]
Ok thanks @orangejenny for all the back and forth. I think this is hopefully ready now. I moved `base_template_minimal.html` into `icds_reports` and resolved duplicate imports. Also, the page didn't need knockout or anything else from `hq` as far as I could tell so I just removed it. Rather add it back when something breaks than leave the bloat in from the beginning.

I actually think there's a chance we can also remove `hq` from the web version of the dashboard too and just add `initial_page_data`, but I'm not ready to sign up for that QA burden for the small reduction in page size yet.

I appreciate getting a better understanding of why you've been pushing so hard for a proper JS dependency system for so long...

File: custom/icds_reports/templates/icds_reports/base_minimal.html
Comment (orangejenny) [inline custom/icds_reports/templates/icds_reports/base_minimal.html]: Didn't notice this earlier - ICDS templates typically don't reference CommCare, right?

Comment (orangejenny) [reply custom/icds_reports/templates/icds_reports/base_minimal.html]: I see now it's overwritten in both child templates, just seems a bit strange.

File: custom/icds_reports/templates/icds_reports/mobile_login.html
Comment (orangejenny) [inline custom/icds_reports/templates/icds_reports/mobile_login.html]: You can remove this now, it gets included by the `javascript_libraries` in the parent.

**Comment by orangejenny** [comment]
@czue This looks in fairly good shape, but I think you've lost analytics handling from `mobile_dashboard.html`.

In general, I think the risk/benefit calculation in optimizing the default libraries is not worth it. Last I looked at the web dashboard it had something like 10MB of js. I'm not sure how stripped-down you're planning to make the mobile version, but the `angular.js` library alone is 1MB. Knockout's main library is 50KB - a completely useless 50KB on the dashboard certainly, but there are likely much bigger fish to fry.

> Rather add it back when something breaks than leave the bloat in from the beginning.

I'd be cautious about this line of thinking beyond knockout (which you might not be doing). Knockout would fail loudly if it were needed and missing, but other js doesn't - things like missing help bubbles or missing analytics could go a while without being noticed.

**Comment by czue** [comment]
Thanks again for the careful review @orangejenny . I think I've addressed all the specific issues you flagged. 

> Knockout's main library is 50KB - a completely useless 50KB on the dashboard certainly, but there are likely much bigger fish to fry.

I think page size is a factor for me, but the bigger factor is comprehensibility. If we leave knockout in, then years from now someone's going to decide to upgrade knockout and then they're going to look everywhere it's used, and then they're not going to be sure whether it's necessary on this page or not and that's going to cause friction where I'd rather it not exist.

Or even not thinking about upgrades, someone's going to decide to decouple knockout and hq from `core_libraries` (which was a reaction i had to seeing that file) and then wonder what level of QA needs to happen on these pages to make that happen (similarly to how I'm feeling about the web dashboard).

So I think if we know we don't need it (which I feel very confident in right now) then I would prefer to leave it out and be explicit about dependencies to the extent possible.

> I'd be cautious about this line of thinking beyond knockout (you're likely already aware of this). Knockout would fail loudly if it were needed and missing, but other js doesn't - things like missing help bubbles or missing analytics could go a while without being noticed.

This point is well taken, though again, I think (:pray: ) that mobile dashboard either won't use these paradigms, or we can introduce them carefully and with testing. I'm tempted to even strip analytics, but there's a lot of web dashboard code that is heavily tangled up with it that we'll probably share...

**Review by orangejenny — APPROVED** [review]

**Comment by czue** [comment]
thanks for riding this out with me @orangejenny !

**Pull request merged**
